### PR TITLE
gaurd against invalid app

### PR DESCRIPTION
**Pull request by Raynos**
reviewers: @jwolski

cc: @sh1mmer

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -53,11 +53,23 @@ var InvalidJoinSourceError = TypedError({
...
+    message: 'Expected `options.app` to be a non-empty string.\n' +
+        'Must specify an app for ringpop to work.\n'
+});
+
 function RingPop(options) {
     if (!(this instanceof RingPop)) {
         return new RingPop(options);
     }
 
+    if (typeof options.app !== 'string' ||
```
Comment (jwolski) [inline index.js]: do `hostPort` too?

Comment (Raynos) [inline index.js]: fixed.

**Comment by Raynos** [comment]
@jwolski Updated this branch with more checks. ALL THE CHECKS.

File: index.js [hunk 2]
```diff
@@ -53,11 +56,59 @@ var InvalidJoinSourceError = TypedError({
...
+    hostPort: null,
+    reason: null
+});
+
 function RingPop(options) {
     if (!(this instanceof RingPop)) {
         return new RingPop(options);
     }
 
+    if (!options) {
```
Comment (jwolski) [inline index.js]: SO MUCH CODEZ.

_required_ options always rubs me the wrong way. It's a contradiction. Rename to `params` now that we're strictly enforcing. Thoughts?

File: index.js [hunk 3]
```diff
@@ -53,11 +56,59 @@ var InvalidJoinSourceError = TypedError({
...
+
+    if (typeof options.app !== 'string' ||
+        options.app.length === 0
+    ) {
+        throw AppRequiredError();
+    }
+
+    var isString = typeof options.hostPort === 'string';
+    var parts = options.hostPort && options.hostPort.split(':');
+    var isColonSeperated = parts && parts.length === 2;
```
Comment (jwolski) [inline index.js]: spelling: separated

File: index.js [hunk 4]
```diff
@@ -53,11 +56,59 @@ var InvalidJoinSourceError = TypedError({
...
+    var isIP = parts && parts[0] && parts[0].match(IP_PATTERN);
```
Comment (jwolski) [inline index.js]: do you need to `match`? How about using just `test()`?

File: index.js (hunk 2 above)
Comment (Raynos) [inline index.js]: `params` is incredibly not idiomatic.

required options are normal in most npm libraries. I understand your feelings but we should stick with `opts` or `options`.

Comment (Raynos) [inline index.js]: > SO MUCH CODEZ

Good user experience requires so much validation boilerplate ;_;

File: index.js (hunk 4 above)
Comment (jwolski) [inline index.js]: asserting on an IP is going to fuck people up. current apps use hostname for host:port

File: index.js (hunk 2 above)
Comment (jwolski) [inline index.js]: ill lean on you for what's more conventional. If you say stick with `options`, let's stick with `options`.

File: index.js (hunk 4 above)
Comment (Raynos) [inline index.js]: @jwolski do we allow hosts ? we assert on IPs in other places.

Comment (jwolski) [inline index.js]: yes, we allow hosts.

**Comment by Raynos** [comment]
@jwolski addressed review feedback.

**Comment by jwolski** [comment]
LGTM

**Comment by Raynos** [comment]
Published v4.1.0

**Pull request merged**